### ISSUE TITLE
Coder acting as reverse proxy, when workspaces are running in separate K8s cluster

### ISSUE THREAD

**Issue by jan94** [coder/coder#9271]
Hi there,
I'm planning to run coder within K8s. For security reasons ( every user of a workspace has a shell and can do what he/she wants ) I would like to run the workspaces themselves in a different, second K8s cluster. As I understood coder itself acts a reverse proxy for the workspaces, therefore I am not sure how this can work when coder and its workspaces are on different clusters...

Could you explain that ? 

How is the reverse proxy embedded in coder proxying the requests in general - in case both coder and the workspaces are in the same cluster ? Is it forwarding the client requests based on the pod IP address of the workspace pod ?

Thanks & BR
Jan

**Comment by kylecarbs** [comment]
In the provider block of your template, you can provide a custom URL:
```tf
provider "coder" {
  url = "http://kubernetes.local.cluster.address
}
```
That way, the second cluster doesn't have to route via the public access URL of your deployment. Coder requires that all workspaces connect to the control plane, and we'll handle the rest after that. Workspaces do not need any port or address exposed.

**Comment by jan94** [comment]
Thank you for the quick response.
Could explain more detailed what you mean by 
'That way, the second cluster doesn't have to route via the public access URL of your deployment.' ?

The second part is clear - only thing that I do not understand is, how does a workspace in the second cluster know which url to contact to register on the control plane ?

**Comment by dcarrion87** [comment]
@jan94 we implement this mechanism exactly and I feel your pain trying to understand how it should all work. The coder server has access to kube creds in the other Kubernetes clusters. We used to have Tailscale proxy within the workspaces namespace to allow coder agents to access the coder without relying on private routing but got rid of it as Coder agent was proxying too much and impacting users. We instead direct route the coder agent access to the coder server at a lower level and overwrite coredns within the kubernetes cluster to reach that server privately. We also tossed the embedded DERP relay in the bin and use our own. Caused users grief whenever we needed to update/patch.
